### PR TITLE
Baggage API: Thread-safety

### PR DESCRIPTION
**Pull request by CodeBlanch**
This is a follow-up to [the discussion](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2284#discussion_r696108973) on  #2284.

## Changes

* The static methods on Baggage which mutate Baggage.Current (SetBaggage, RemoveBaggage, & ClearBaggage) are now thread-safe within the scope of an ExecutionContext.
* The instance methods on Baggage no longer mutate Baggage.Current.
 
## Breaking behavior changes

Previously if you called SetBaggage, RemoveBaggage, or ClearBaggage on a Baggage instance it would mutate Baggage.Current. For example:

```csharp
var baggage = Baggage.Current;
baggage.SetBaggage(K1, V1);

Baggage.Current.SetBaggage(K2, V2);
```

Now you have to explicitly set back the instance. Like this:

```csharp
var baggage = Baggage.Current;
baggage.SetBaggage(K1, V1);
Baggage.Current = baggage;

Baggage.Current = Baggage.Current.SetBaggage(K2, V2);
```

If you use the statics, this is done automatically:

```csharp
Baggage.SetBaggage(K1, V1);
Baggage.SetBaggage(K2, V2);
```

The reason for the change is Baggage is a `readonly struct` so when you do Baggage.Current you get a copy. You can write code like this:

```csharp
var currentBaggage = Baggage.Current;
currentBaggage.SetBaggage(K1, V1); // User probably wants this to become Current

var detachedBaggage = Baggage.Create();
detachedBaggage.SetBaggage(K2, V2); // What does the user want in this case? Mutating Current might be unexpected
```

Basically when the instance SetBaggage executes it doesn't know if it is current or not. I thought about managing a flag on Baggage so it could know if it was attached or not, but it is readonly so there's no (safe) way to update the flag. We could compare it back to Current, but another thread could have already changed it. I decided the best option was to always treat it as detached and allow the user to make the decision about whether or not it should become the "Current" version or not. I think this makes the API more consistent and usable but it will break anyone reliant on this behavior today.

Alternatively, we could [Obsolete] Baggage and introduce something new with a contract.

## TODOs

* [X] `CHANGELOG.md` updated for non-trivial changes

### CODE REVIEW

**Comment by CodeBlanch** [comment]
Created as a draft to get feedback on the breaking behavior change. If we're OK with this, I'll update changelog with a note and convert to a real PR.

/cc @cijothomas @alanwest @reyang

File: src/OpenTelemetry.Api/Baggage.cs [hunk 1]
```diff
@@ -46,21 +46,34 @@ internal Baggage(Dictionary<string, string> baggage)
...
         /// </summary>
+        /// <remarks>
+        /// Note: <see cref="Current"/> returns a copy of the current Baggage.
+        /// Changes to the returned instance will not automatically be reflected
+        /// back on the <see cref="Current"/> version. To update the <see
+        /// cref="Current"/> version either set <see cref="Current"/> to a new
+        /// instance or use one of the static methods that target <see
+        /// cref="Current"/> as the default source. Examples:
+        /// <code>
+        /// Baggage.SetBaggage("newKey1", "newValue1"); // Updates Baggage.Current with 'newkey1'
```
Comment (reyang) [inline src/OpenTelemetry.Api/Baggage.cs:57]: nit:

```suggestion
        /// Baggage.SetBaggage("newKey1", "newValue1"); // Updates Baggage.Current with 'newKey1'
```

File: src/OpenTelemetry.Api/Baggage.cs [hunk 2]
```diff
@@ -46,21 +46,34 @@ internal Baggage(Dictionary<string, string> baggage)
         /// <summary>
         /// Gets or sets the current <see cref="Baggage"/>.
         /// </summary>
+        /// <remarks>
+        /// Note: <see cref="Current"/> returns a copy of the current Baggage.
```
Comment (reyang) [inline src/OpenTelemetry.Api/Baggage.cs:50]: minor: I personally find it confusing when we use the word "copy" since it could mean many different things - shallow copy, deep copy, copy by ref, shadow copy, copy-on-write, etc. I wonder if the word "fork" would give more clarity here - e.g. "returns a forked version of the Baggage" and "any updates on the forked version will not affect the Current version". I don't feel too strong about it though.

File: test/OpenTelemetry.Tests/BaggageTests.cs [hunk 3]
```diff
@@ -298,5 +300,18 @@ static async Task InnerTask()
                 // key2 & key3 changes don't flow backward automatically
             }
         }
+
+        [Fact]
+        public void ThreadSafetyTest()
+        {
+            Baggage.SetBaggage("root", "root");
```
Comment (reyang) [inline test/OpenTelemetry.Tests/BaggageTests.cs:307]: minor:

```suggestion
            Baggage.SetBaggage("rootKey", "rootValue");
```

**Review by reyang — APPROVED** [review]
LGTM, it is not a breaking change from the signature perspective.

It does change the behavior which _might_ break the users if they rely on certain behavior, but I feel that if we leave it as is (which is a bit confusing I guess?) we probably will end up surprising/confusing more users moving forward.

File: src/OpenTelemetry.Api/Baggage.cs [hunk 4]
```diff
@@ -46,21 +46,34 @@ internal Baggage(Dictionary<string, string> baggage)
...
+        /// Baggage.SetBaggage("newKey2", "newValue2"); // Updates Baggage.Current with 'newkey2'
+        /// </code>
+        /// Or:
+        /// <code>
+        /// var baggageCopy = Baggage.Current;
+        /// Baggage.SetBaggage("newKey1", "newValue1"); // Updates Baggage.Current with 'newkey1'
+        /// var newBaggage = baggageCopy
+        ///     .SetBaggage("newKey2", "newValue2");
+        ///     .SetBaggage("newKey3", "newValue3");
+        /// /* Sets Baggage.Current to 'newBaggage' which will override any
```
Comment (reyang) [inline src/OpenTelemetry.Api/Baggage.cs:67]: nit: maybe change this to line comments (mixing the two types of comments is a bit weird)

```suggestion
        /// // Sets Baggage.Current to 'newBaggage' which will override any
```

File: src/OpenTelemetry.Api/Baggage.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Api/Baggage.cs]: I tweaked it to use "fork" LMK what you think.

Comment (reyang) [reply src/OpenTelemetry.Api/Baggage.cs]: Looks great, thanks!

File: src/OpenTelemetry.Api/Baggage.cs [hunk 5]
```diff
@@ -46,21 +46,34 @@ internal Baggage(Dictionary<string, string> baggage)
...
         /// Gets or sets the current <see cref="Baggage"/>.
         /// </summary>
+        /// <remarks>
+        /// Note: <see cref="Current"/> returns a forked version of the current
+        /// Baggage. Changes to the forked version will not automatically be
+        /// reflected back on <see cref="Current"/>. To update <see
+        /// cref="Current"/> either set <see cref="Current"/> to a new instance
+        /// or use one of the static methods that target <see cref="Current"/>
+        /// as the default source. Examples:
+        /// <code>
```
Comment (cijothomas) [inline src/OpenTelemetry.Api/Baggage.cs:56]: can you reverse the order of the example codes below, so that it matches the Note: section.

**Review by cijothomas — APPROVED** [review]
LGTM.
Suggested to modify the readme to guide users to use static (Baggage.SetBaggage).

File: src/OpenTelemetry.Api/Baggage.cs (hunk 5 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Api/Baggage.cs:57]: I reversed your request to reverse the code examples and instead reversed the notes. I wanted to have the preferred approach (static methods) first. I also updated the README to the safer static API.

Comment (reyang) [reply src/OpenTelemetry.Api/Baggage.cs:57]: > I reversed your request to reverse the code examples and instead reversed the notes.

This makes my day 😵😆

Comment (CodeBlanch) [reply src/OpenTelemetry.Api/Baggage.cs:57]: Ya had some fun with that one 😄 Also very excited about the pun opportunities for my next PR: "Lost Baggage"

**Review by alanwest — APPROVED** [review]

File: src/OpenTelemetry.Api/README.md [hunk 6]
```diff
@@ -85,26 +85,25 @@ OpenTelemetry SDK ships a BaggagePropagator and enables it by default.
 
 ```csharp
 // Use Baggage.Current to get all the key/value pairs present in Baggage
-foreach (var item in Baggage.Current)
+foreach (var item in Baggage.GetBaggage())
```
Comment (cijothomas) [inline src/OpenTelemetry.Api/README.md:88]: the line above has a comment referring to Baggage.Current.

Comment (CodeBlanch) [reply src/OpenTelemetry.Api/README.md:88]: Fixed
